src/java.base/share/classes/java/util/concurrent/Joiners.java

```diff
@@ -0,0 +1,220 @@
+ * by Oracle in the LICENSE file that accompanied this code.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+package java.util.concurrent;
+
+import java.lang.invoke.MethodHandles;
+import java.lang.invoke.VarHandle;
+import java.util.ArrayList;
+import java.util.Comparator;
+import java.util.List;
+import java.util.NoSuchElementException;
+import java.util.Objects;
+import java.util.concurrent.StructuredTaskScope.Joiner;
+import java.util.concurrent.StructuredTaskScope.Subtask;
+import java.util.function.Predicate;
+import java.util.stream.Stream;
+import jdk.internal.invoke.MhUtil;
+
+/**
+ * Built-in StructuredTaskScope.Joiner implementations.
+ */
+class Joiners {
+    private Joiners() { }
+
+    /**
+     * Throws IllegalArgumentException if the subtask is not in the UNAVAILABLE state.
+     */
+    private static void ensureUnavailable(Subtask<?> subtask) {
+        if (subtask.state() != Subtask.State.UNAVAILABLE) {
+            throw new IllegalArgumentException("Subtask not in UNAVAILABLE state");
+        }
+    }
+
+    /**
+     * Throws IllegalArgumentException if the subtask has not completed.
+     */
+    private static Subtask.State ensureCompleted(Subtask<?> subtask) {
+        Subtask.State state = subtask.state();
+        if (state == Subtask.State.UNAVAILABLE) {
+            throw new IllegalArgumentException("Subtask has not completed");
+        }
+        return state;
+    }
+
+    /**
+     * A joiner that returns a stream of all subtasks when all subtasks complete
+     * successfully. Cancels the scope if any subtask fails.
+     */
+    static final class AllSuccessful<T> implements Joiner<T, Stream<Subtask<T>>> {
+        private static final VarHandle FIRST_EXCEPTION =
+                MhUtil.findVarHandle(MethodHandles.lookup(), "firstException", Throwable.class);
+
+        // list of forked subtasks, only accessed by owner thread
+        private final List<Subtask<T>> subtasks = new ArrayList<>();
+
+        private volatile Throwable firstException;
+
+        @Override
+        public boolean onFork(Subtask<? extends T> subtask) {
+            ensureUnavailable(subtask);
+            @SuppressWarnings("unchecked")
+            var s = (Subtask<T>) subtask;
+            subtasks.add(s);
```

Comment:
```suggestion
            subtasks.add((Subtask<T>) subtask);
```
The local variable s is only used once and can be simplified

Reply:
No, it has to be a local to avoid broadening the use of SW.